aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -16,7 +16,7 @@ namespace meta {
 TORCH_META_FUNC2(add, Tensor) (
   const Tensor& self, const Tensor& other, const Scalar& alpha
 ) {
-  build_binary_op(maybe_get_output(), self, other);
+  build_borrowing_binary_op(maybe_get_output(), self, other);
```

Comment:
On the migration plan- some options sound like:

(1) Don't bother including both the borrowing and non-borrowing API's. Immediately opt all TensorIterator kernels to use borrowing

(2) Hitch this to structured kernels- all TI ops that are ported to structured will automatically use the borrowing interface

(3) Port kernels one-by-one to use borrowing (like in this PR).


Maybe (2) is a reasonable middle ground? (3) seems like too much work, but I could understand not doing (1), if we're worried about the API not getting used properly since we don't have a borrow checker

Reply:
I'm not sure how to accomplish (2) concretely -- the code block we're commenting on is a structured kernel, and it required a manual one-by-one code change. How would you fix it?

Reply:
One way is that unstructured and structured TI kernels have slightly different code paths. Structured kernels call into `TensorIteratorBase::build_binary_op` (or similar, for unary, etc), while unstructured kernels call `TensorIterator::binary_op` ([example](https://github.com/pytorch/pytorch/blob/f3367f917e916a9f57461089889f4f4ddbb16d20/aten/src/ATen/native/BinaryOps.cpp#L161)). Right now `binary_op` is written in terms if `build_binary_op`, but we could update it so that it doesn't use borrowing and `build_binary_op` does.

I'm not 100% sure on the benefit of only applying this to structured kernels though, aside from it being less of a "big bang" change. I doubt everyone who ports an op to be structured will know to think carefully about the borrowing logic and whether or not it would break for their op.

Reply:
> I doubt everyone who ports an op to be structured will know to think carefully about the borrowing logic and whether or not it would break for their op.

Sounds like it's better to be explicit about what's going on, then.